File: 백준/Silver/9012. 괄호/괄호.cc
#include <bits/stdc++.h>
using namespace std;

int main() {
	int t; 
	string s;
	cin >> t;

	while (t--) {
		stack<char> st;
		int flag = 0;
		cin >> s;

		for (auto c : s) {
			if (st.empty()) {
				if (c == ')') {
					flag = 1;
					break;
				}
				st.push(c);
			} 
			else {
				if (st.top() != c) {
					st.pop();
				} else {
					st.push(c);
				}
			}
		}
		
		if (flag == 1) cout << "NO\n";
		else if (st.size() != 0) cout << "NO\n";
		else cout << "YES\n";
	}
}

File: 백준/Silver/9012. 괄호/괄호.py

t = int(input())

for _ in range(t):
    stack = []
    ps = list(input())

    for c in ps:
        if not stack:
            stack.append(c)
        elif stack[-1] == '(' and c == ')':
            stack.pop()
        else:
            stack.append(c)
    if not stack: print('YES')
    else: print('NO')


